SpatialGDK/Source/SpatialGDK/Private/Interop/ActorSystem.cpp

```diff
@@ -0,0 +1,1821 @@
+				   EntityId, ComponentId);
+			return;
+		}
+
+		ApplyComponentUpdate(ComponentId, Update, *TargetObject, *Channel, /* bIsHandover */ true);
+	}
+	else
+	{
+		UE_LOG(LogActorSystem, Verbose,
+			   TEXT("Entity: %d Component: %d - Skipping because it's an empty component update from an RPC component. (most likely as a "
+					"result of gaining authority)"),
+			   EntityId, ComponentId);
+	}
+}
+
+void ActorSystem::ComponentRemoved(const Worker_EntityId EntityId, const Worker_ComponentId ComponentId) const
+{
+	// We want to do nothing for RemoveComponent ops for which we never received a corresponding
+	// AddComponent op. This can happen because of the worker SDK generating a RemoveComponent op
+	// when a worker receives authority over a component without having already received the
+	// AddComponent op. The generation is a known part of the worker SDK we need to tolerate for
+	// enabling dynamic components, and having authority ACL entries without having the component
+	// data present on an entity is permitted as part of our Unreal dynamic component implementation.
+	if (!EntityHasComponent(EntityId, ComponentId))
+	{
+		return;
+	}
+
+	if (AActor* Actor = Cast<AActor>(NetDriver->PackageMap->GetObjectFromEntityId(EntityId).Get()))
+	{
+		FUnrealObjectRef ObjectRef(EntityId, ComponentId);
+		if (ComponentId == SpatialConstants::DORMANT_COMPONENT_ID)
+		{
+			GetOrRecreateChannelForDomantActor(Actor, EntityId);
+		}
+		else if (UObject* Object = NetDriver->PackageMap->GetObjectFromUnrealObjectRef(ObjectRef).Get())
+		{
+			if (USpatialActorChannel* Channel = NetDriver->GetActorChannelByEntityId(EntityId))
+			{
+				TWeakObjectPtr<UObject> WeakPtr(Object);
+				Channel->OnSubobjectDeleted(ObjectRef, Object, WeakPtr);
+
+				Actor->OnSubobjectDestroyFromReplication(Object);
+
+				Object->PreDestroyFromReplication();
+				Object->MarkPendingKill();
+
+				NetDriver->PackageMap->RemoveSubobject(FUnrealObjectRef(EntityId, ComponentId));
+			}
+		}
+	}
+}
+
+void ActorSystem::EntityAdded(const Worker_EntityId EntityId)
+{
+	ReceiveActor(EntityId);
+	for (const auto& AuthoritativeComponent : SubView->GetView()[EntityId].Authority)
+	{
+		AuthorityGained(EntityId, AuthoritativeComponent);
+	}
+}
+
+void ActorSystem::EntityRemoved(const Worker_EntityId EntityId)
+{
+	// SCOPE_CYCLE_COUNTER(STAT_ReceiverRemoveEntity);
+
+	RemoveActor(EntityId);
+
+	// Stop tracking if the entity was deleted as a result of deleting the actor during creation.
+	// This assumes that authority will be gained before interest is gained and lost.
+	const int32 RetiredActorIndex = EntitiesToRetireOnAuthorityGain.IndexOfByPredicate([EntityId](const DeferredRetire& Retire) {
+		return EntityId == Retire.EntityId;
+	});
+	if (RetiredActorIndex != INDEX_NONE)
+	{
+		EntitiesToRetireOnAuthorityGain.RemoveAtSwap(RetiredActorIndex);
+	}
+
+	if (NetDriver->IsServer())
```

Comment:
This should live outside the actor system, since it is used to detect client disconnection by looking at the system entities.